ref(normalization): Remove StoreProcessor

iker-barriocanal opened the pull request:
This PR removes the `StoreProcessor`, bringing us one step closer to removing the store normalizer entirely.

The biggest difference is that normalization of default attributes is now in `event::normalize` and runs in all relays, not just processing.

### Design choices

- There's a `StoreConfig` struct in the c-abi. Setting the configuration from librelay should happen with a single object and Relay uses 3 (2 for validation and 1 for normalization), so we can't get rid of it in the current setting.
- The normalization config has a `is_last_normalize`. Relay needs to know whether it's the last normalization step to e.g. remove unknown fields not to break future compatibility. This flag is currently enabled in processing relays and librelay calls, and it's disabled by default.

### Future work
- Move `normalize::mod` contents to their appropriate place. Currently, it contains a few types, a couple of functions used in other places, and tests validating the functionality in other modules.
- Make running normalization configurable, and only run it in PoP relays.

#skip-changelog

iker-barriocanal committed:
ref(normalization): Remove StoreNormalizeProcessor

olksdr reviewed (commented):
@iker-barriocanal  are those failing integration test just flaky or failing for a reason? 
Could you, please, have a look into them before someone take a closer look to this PR? 
Thanks!

iker-barriocanal committed:
Merge branch 'master' into iker/ref/rm-storenormalizeproc

iker-barriocanal committed:
fix tests

iker-barriocanal committed:
Merge branch 'master' into iker/ref/rm-storenormalizeproc

iker-barriocanal commented on tests/integration/test_store.py:
```diff
@@ -263,20 +263,6 @@ def store_event():
     store_count.acquire(timeout=2)
 
 
-def test_store_not_normalized(mini_sentry, relay):
```
The test is stale as normalization runs for all relays, including the default attributes (like `key_id`).

iker-barriocanal commented on tests/integration/fixtures/security_report/csp_chrome_blocked_asset.no_processing.output.json:
```diff
@@ -23,12 +23,18 @@
   },
   "culprit": "style-src cdn.example.com",
   "environment": "production",
+  "grouping_config": {
```
These snapshot updates are the result of normalizing default attributes in regular normalization.

iker-barriocanal committed:
Merge branch 'master' into iker/ref/rm-storenormalizeproc

Dav1dde reviewed (approved):
As discussed offline, this is just an intermediate step to get rid of the `StoreProcessor` in the coming days which will cleanup a bunch of additional config and code which might be confusing for now (e.g. `event::store` and other callsites and misleading config).

iker-barriocanal committed:
remove storeprocessor

iker-barriocanal committed:
Merge branch 'master' into iker/ref/rm-storenormalizeproc

iker-barriocanal commented on relay-event-normalization/src/event.rs:
```diff
@@ -156,6 +212,18 @@ pub fn normalize_event(event: &mut Annotated<Event>, config: &NormalizationConfi
+    if config.is_last_normalize && config.remove_other.unwrap_or(!is_renormalize) {
+        // Remove unknown attributes at every level
+        let _ =
+            remove_other::RemoveOtherProcessor.process_event(event, meta, ProcessingState::root());
```
These two processors don't return `ProcessingResult` errors that drop events, so it's safe to ignore the result.

Dav1dde replied:
Would be good to add this comment to the code

iker-barriocanal replied:
I find the comment redundant -- if you're reading it, you already see the result being ignored.

iker-barriocanal commented on relay-cabi/src/processing.rs:
```diff
@@ -86,32 +206,31 @@ pub unsafe extern "C" fn relay_valid_platforms(size_out: *mut usize) -> *const R
 ) -> RelayStr {
-    let processor = normalizer as *mut StoreProcessor;
+    let config = config as *mut StoreConfig;
+    let config = (*config).this();
```
Making this in one line is less readable:
```rs
(*(config as *mut StoreConfig)).this()
```

iker-barriocanal committed:
fix + todo

Dav1dde commented on relay-event-normalization/src/event.rs:
```diff
@@ -201,6 +212,18 @@ pub fn normalize_event(event: &mut Annotated<Event>, config: &NormalizationConfi
             trimming::TrimmingProcessor::new().process_event(event, meta, ProcessingState::root());
     }
+
+    if config.is_last_normalize && config.remove_other.unwrap_or(!is_renormalize) {
```
I find the behaviour strange if I explicitly state `remove_other: Some(true)` and then it does not run because it depends on `is_last_normalize` (additionally this is not documented).

Possibly include the check only in the `unwrap_or`.

iker-barriocanal replied:
Using a single option in [59c0c95](https://github.com/getsentry/relay/pull/3097/commits/59c0c950929b63e64998e8ee817a21e7812b8dba).

Dav1dde commented on relay-event-normalization/src/event.rs:
```diff
@@ -109,6 +115,9 @@ pub struct NormalizationConfig<'a> {
     pub is_renormalize: bool,
 
+    /// Overrides the default flag for other removal.
+    pub remove_other: Option<bool>,
```
Maybe in a follow-up we can start removing these optionals?

The `StoreConfig` in the C-ABI can still have the optionals, but the NormalizationConfig doesn't have to have it.

iker-barriocanal replied:
Done in [59c0c95](https://github.com/getsentry/relay/pull/3097/commits/59c0c950929b63e64998e8ee817a21e7812b8dba).

Dav1dde commented on relay-cabi/src/processing.rs:
```diff
@@ -9,29 +9,149 @@ use std::os::raw::c_char;
+/// Configuration for the store step -- validation and normalization.
+#[derive(Serialize, Deserialize, Debug, Default)]
+#[serde(default)]
+pub struct StoreConfig {
```
I think we want to call this `StoreNormalizer` or wrap the config in a type called `StoreNormalizer` so we have still a consistent naming with the C-ABI functions.

Dav1dde commented on relay-cabi/src/processing.rs:
```diff
@@ -9,29 +9,149 @@ use std::os::raw::c_char;
+
+impl StoreConfig {
+    /// Helper method to parse *mut StoreConfig -> &StoreConfig
+    pub fn this(&self) -> &Self {
```
```suggestion
   fn this(&self) -> &Self {
```

Looks like this doesn't need to be public

Dav1dde commented on relay-event-normalization/src/event.rs:
```diff
@@ -31,14 +31,20 @@ use crate::normalize::request;
+    /// Whether it's the last normalization step.
+    ///
+    /// The last normalization run performs a few additional steps that
+    /// shouldn't happen in previous runs, like removing unknown attributes.
```
Not quite true, it also needs `remove_other` to be enabled.

I think we can kill it in favor of having specialized bools like `remove_other` and `emit_events`

jjbayer replied:
> I think we can kill it in favor of having specialized bools

+1, `is_last_normalize` could be misleading because the final normalization step actually happens in sentry.

iker-barriocanal replied:
Killed the config in [59c0c95](https://github.com/getsentry/relay/pull/3097/commits/59c0c950929b63e64998e8ee817a21e7812b8dba).

jjbayer commented on relay-server/src/services/processor.rs:
Should we rename `light_normalize_event` to `normalize_event` now?

iker-barriocanal replied:
Renamed most references in [59c0c95](https://github.com/getsentry/relay/pull/3097/commits/59c0c950929b63e64998e8ee817a21e7812b8dba). There are a couple left that I'll address in a follow-up as they'll require a bit more work than just renaming.

jjbayer commented on relay-event-normalization/src/event.rs:
```diff
@@ -894,6 +1019,210 @@ fn normalize_breakdowns(event: &mut Event, breakdowns_config: Option<&Breakdowns
+/// Returns `true` if the given platform string is a known platform identifier.
+///
+/// See [`VALID_PLATFORMS`] for a list of all known platforms.
+pub fn is_valid_platform(platform: &str) -> bool {
```
nit: Can we remove this to the top, because it's public?

iker-barriocanal replied:
All definitions above are also public. Where would you move it?

iker-barriocanal committed:
align naming on cabi and more accurate docs

jjbayer commented on relay-cabi/src/processing.rs:
```diff
@@ -80,38 +200,37 @@ pub unsafe extern "C" fn relay_valid_platforms(size_out: *mut usize) -> *const R
-    let config: StoreConfig = serde_json::from_str((*config).as_str())?;
-    let normalizer = StoreProcessor::new(config);
-    Box::into_raw(Box::new(normalizer)) as *mut RelayStoreNormalizer
+    let config: StoreNormalizer = serde_json::from_str((*config).as_str())?;
```
```suggestion
    let normalizer: StoreNormalizer = serde_json::from_str((*config).as_str())?;
```

jjbayer replied:
same for functions below

iker-barriocanal replied:
Done in [59c0c95](https://github.com/getsentry/relay/pull/3097/commits/59c0c950929b63e64998e8ee817a21e7812b8dba).

jjbayer commented on relay-cabi/src/processing.rs:
```diff
@@ -9,24 +9,144 @@ use std::os::raw::c_char;
+    /// To normalize spans in [`normalize_event`], `is_renormalize` must
+    /// be disabled _and_ `normalize_spans` enabled.
+    pub normalize_spans: bool,
+}
```
Is this a 1:1 copy of the old `StoreConfig`, or were there changes? Might be worth leaving the class definition in the old location to make the diff easier to read.

iker-barriocanal replied:
Yep, it's a 1:1 copy

Dav1dde replied:
I like that is moved, Relay itself doesn't need it anymore.

jjbayer replied:
I like that it's moved too! My comment was about diff readability. But if there's no diff then it's fine.

iker-barriocanal committed:
feedback

- remove is_last_normalize in favor of per-step configuration
- remove some references to light normalization

jjbayer commented on relay-cabi/src/processing.rs:
```diff
@@ -127,8 +246,16 @@ pub unsafe extern "C" fn relay_store_normalizer_normalize_event(
+    let is_renormalize = config.is_renormalize.unwrap_or(false);
+
     let normalization_config = NormalizationConfig {
+        project_id: config.project_id,
+        client: config.client.clone(),
+        protocol_version: config.protocol_version.clone(),
+        key_id: config.key_id.clone(),
+        grouping_config: config.grouping_config.clone(),
```
Do these have to be clone?

iker-barriocanal replied:
Yep, `NormalizationConfig` expects `String`s (and `serde_json::Value` for `grouping_config`). I'll look into the config to see if it needs to own the data.

jjbayer commented on relay-cabi/src/processing.rs:
```diff
@@ -137,7 +264,9 @@ pub unsafe extern "C" fn relay_store_normalizer_normalize_event(
-        is_renormalize: config.is_renormalize.unwrap_or(false),
+        is_renormalize,
+        remove_other: config.remove_other.unwrap_or(!is_renormalize),
+        emit_error_events: !is_renormalize,
```
```suggestion
        emit_event_errors: !is_renormalize,
```

jjbayer reviewed (approved):
There's a small chance that this breaks things, because processing steps running between light normalization and store normalization could in theory modify the event in ways that were previously caught by store normalization. But it's a risk worth taking IMO.

iker-barriocanal committed:
emit_error_events -> emit_event_errors

iker-barriocanal committed:
fix missing rename
